src/main/java/gololang/LazyList.java

```diff
@@ -0,0 +1,396 @@
+ * is represented by a closure that is evaluated only if needed.
+ * The value is memoized, so that the closure representing the tail 
+ * is evaluated only once. 
+ *
+ * Since the tail closure will be called at most once, and we can't 
+ * guarantee when, or even if, it will be called, this closure must be 
+ * a pure, side-effect free, function.
+ */
+public class LazyList implements Collection<Object>, HeadTail<Object> {
+
+  /**
+   * Represents the empty list.
+   */
+  public static final LazyList EMPTY = new LazyList(null, null) {
+    @Override
+    public boolean equals(Object other) {
+      return other == this;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(null, null);
+    }
+
+    @Override
+    public boolean isEmpty() { 
+      return true;
+    }
+
+    @Override
+    public int size() {
+      return 0;
+    }
+
+    @Override
+    public LazyList tail() {
+      return this;
+    }
+
+    @Override
+    public String toString() {
+      return "LazyList.EMPTY";
+    }
+  };
+
+  /**
+   * Iterator over a {@code LazyList}.
+   */
+  class LazyListIterator implements Iterator<Object> {
+
+    private LazyList list;
+
+    LazyListIterator(LazyList aLazyList) {
+      this.list = aLazyList;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return !list.isEmpty();
+    }
+
+    @Override
+    public Object next() {
+      Object h = list.head();
+      list = list.tail();
+      return h;
+    }
+
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException(
+          "remove() is not supported on a lazy list iterator");
+    }
+  }
+
+  private final Object head;
+  private final FunctionReference tail;
+  private LazyList memoTail = null;
+
+  /**
+   * Create a new list from the head and tail values.
+   *
+   * @param head the first value of the list.
+   * @param tail a {@code FunctionReference} that returns a LazyList when invoked.
+   * @return a new {@code LazyList}
+   */
+  public static LazyList cons(Object head, FunctionReference tail) {
+    if (tail == null) {
+      throw new IllegalArgumentException("Use the empty list instead of null as the last element of a LazyList");
+    }
+    return new LazyList(head, tail);
+  }
+
+  private LazyList(Object head, FunctionReference tail) {
+    this.head = head;
+    this.tail = tail;
+  }
+
+  /**
+   * Gets the first element of the list (its head).
+   *
+   * @return an {@code Object}, or {@code null} if the list is empty.
+   */
+  public Object head() {
+    return this.head;
+  }
+
+  /**
+   * Gets the rest of the list (its tail).
+   *
+   * @return a {@code LazyList}, or {@code EMPTY} if the list is empty,
+   * contains only one value, or if the closure failed.
+   */
+  public LazyList tail() {
+    if (memoTail == null) {
+      try {
+        memoTail = (LazyList) (this.tail.invoke());
+      } catch (Throwable e) {
+        memoTail = EMPTY;
```

Comment:
There are many cases where this exception can't be properly handled by the user (e.g. when performing a `map` on the list). In this case, it seems more practical to simply terminate the list. This is semantically correct w.r.t. the lazyness of the list. If the next function can't perform, then the list is empty. The `tail` closure is never called twice, since the result is cached. Since the functions representing the next value *must* be pure functions with no parameters, it *must* always return the same value (or fail to run each time).